[BE] [Team-08] ToDo-list 최종 구현

jihye-woo opened the pull request:
안녕하세요!! @Polynomeer(August) @jihye-woo(Woody) 입니다!!😀

### 개요

🔗 [API 배포 링크](http://54.180.218.17:8080/api/cards)
🔗 [API Document](https://github.com/dahun-lee-daji/todo-list/wiki/API-Document)

### 구현 내용
#### 1. Card에 대한 CRUD 구현
- CardRepository에서 Card를 읽어올 때, MySQL의 TINYINT와 Java의 boolean 타입이 맞지 않아서 직접 쿼리 작성

#### 2. History 조회 구현

#### 3. History 생성 로직 구현
- http 요청에 따라 히스토리 생성
- HistoryService에서 각 action에 대한 generate 메소드를 가지도록 구현
- Move와 Update는 소속 columnId의 변경 유무로 구분


감사합니다!!✨

jihye-woo committed:
feat: Add Entity classes (#6)

- Card, History, User 클래스 구현
- 각 엔티티에 getters 추가
- DateTimeUtils를 통해 LocalDateTime을 포멧팅하도록 함

jihye-woo committed:
feat: Add API Controllers (#6)

- 각 엔티티에 대한 RestController를 구현하여 json으로 보낼 수 있도록 함
- 더미 데이터를 직접 생성하여 리스트에 담아 전달

jihye-woo committed:
feat: Add update and delete methods for Card (#6)

- @PutMapping으로 ApiCardController::update 메소드 정의
- @DeleteMapping으로 ApiCardController::delete 메소드 정의
- "/api/cards" url 수정

jihye-woo committed:
feat: Add Item entity class and controller (#6)

- @GetMapping으로 ApiItemController::list를 추가하여 간단한 item 더미 데이터 리턴
- Item 엔티티 클래스 추가

jihye-woo committed:
build: Change JDK version from 11 to 1.8

jihye-woo committed:
feat: Modify fields of entity classes (#6)

- db 테이블 스키마 변경에 따른 엔티티 클래스 필드 수정
- 컨트롤러에서 더미 데이터 생성 코드 수정

jihye-woo committed:
build: Remove test dependencies (#6)

- Junit과 assertJ 의존성 제거
- jar 파일 빌드 시, 에러 발생을 방지

jihye-woo committed:
refactor: Modify type and naming of entities (#6)

- Item 클래스 명을 Column으로 모두 변경
- 모든 프리미티브 타입을 레퍼 클래스 타입으로 변경
- Card::isApp 메소드를 getIsApp으로 변경하여 응답 데이터 포멧을 맞춤

jihye-woo committed:
feat: Add DTO and adapt to controllers (#16)

- Card와 History에 대한 DTO 생성
- 각 컨트롤러에 DTO를 적용시켜 필요한 데이터만 클라이언트로 전송

jihye-woo committed:
feat: Add mockup to update and delete (#16)

- ApiCardController::update에서 전달받은 CardDto 정보를 그대로 리턴
- ApiCardController::delete에서 전달받은 cardId를 포함한 메시지를 리턴
- DateTimeUtils::reformatByPattern를 추가 (String to LocalDateTime)

jihye-woo committed:
feat: Add mockup to card create (#16)

- ApiCardController::create에 전달받은 CardDto 정보를 그대로 리턴

jihye-woo committed:
refactor: Remove unused methods in controllers (#16)

- ApiCardController::list 삭제
- ApiCardController::getCard 삭제
- ApiColumnController::list 삭제

jihye-woo committed:
build: Add Spring Boot custom banner (#16)

- resources 디렉토리에 banner.txt 추가

jihye-woo committed:
chore: Add various dummy data in histories (#16)

- ApiHistoryController::list에 더미 데이터 추가
- 다른 시간대의 히스토리 데이터를 7개 생성

jihye-woo committed:
feat: Add Services and Repositories (#17)

- 각 엔티티 PK에 @Id 어노테이션 추가
- 각 엔티티에 대한 서비스 및 레퍼지토리 생성 및 의존성 주입
- 스프링 부트 웹 어플리케이션에 대한 계층적 구조만 작성(DB 연동 X)

jihye-woo committed:
feat: Add Card create logic (#17)

- CardService::create에서 CardRepository의 save 메소드 이용
- CardDto::toEntity 추가하여 CardRepository::save 사용할 수 있도록 함
- Card 엔티티에 columnId 필드 추가하여 소속 column을 알 수 있도록 함

jihye-woo committed:
chore: Update dummy data (#17)

- ApiCardController::list 및 더미 데이터 추가
- ApiHistoryController::list에서 더미 데이터 수
  - 모든 액션에 대하여 to에 반드시 데이터가 들어가도록 수정 (from은 nullable)

jihye-woo committed:
feat: Add card listing logic (#17)

- /api/cards에 대한 로직을 ApiCardController::list에 구현
- CardDto::of를 추가하여 Card 엔티티를 CardDto로 맵핑
- CardRepository::findAll에 쿼리문을 직접 작성하여 boolean 타입으로 변환하여 isApp을 가져옴
- 각 엔티티의 필드명을 DB 칼럼명에 맞게 수정
- 불필요한 ApplicationConfig.java 삭제

jihye-woo committed:
feat: Add card create logic (#17)

- POST /api/cards 요청에 대해 카드 생성 로직 구현
- Aggregate 문제로 column 관련 클래스 모두 제거

jihye-woo committed:
feat: Add card delete logic (#17)

- DELETE /api/cards/{cardId} 요청에 대해 카드 삭제 로직 구현

jihye-woo committed:
feat: Add card update logic (#17)

- PUT /api/cards/{cardId} 요청에 대해 카드 수정 로직 구현
- CardRepository::findById를 재정의하여 isApp 타입 변환
- Card::update 메소드를 추가

jihye-woo committed:
feat: Add history read logic (#17)

- GET /api/histories 요청에 대해 히스토리 조회 로직 구현
- HistoryRepository::findAll 재정의해서 List 리턴
- HistoryDto::of 추가하여 엔티티를 DTO로 변환
- 테이블 칼럼명에 맞게 엔티티 필드명 수정

jihye-woo committed:
feat: Add @Table to handle lowercase (#17)

- AWS의 mysql 설정은 대소문자를 구분하지만 로컬환경에서는 구별하지 않으므로 호환성을 위해 수정

jihye-woo committed:
feat: Delete cid in History (#17)

- History의 cid(card id)로 인해 외래키 제약조건을 위반하여 카드가 삭제되지 않는 문제 때문에 제거

jihye-woo committed:
feat: Add history generation logic for add action (#25)

- HistoryService::generateAdd를 추가하여 add 액션에 대한 히스토리 생성
- ApiCardController::create에서 generateAdd로 히스토리 생성
- DateTimeUtils를 util 패키지로 이동

jihye-woo committed:
feat: Add history generation logic for remove, update, move action (#25)

- HistoryService::generateUpdate, generateRemove, generateMove를 추가하여 update, remove, move 액션에 대한 히스토리 생성 및 저장
- ApiCardController에서 HistoryService를 통해 update, remove, move 액션에 대한 히스토리 생성 메소드 호출

wheejuni commented on backend/src/main/java/com/team08/todolist/controller/ApiCardController.java:
```diff
@@ -0,0 +1,63 @@
+    @PutMapping("/{cardId}")
+    public CardDto update(@PathVariable Long cardId, CardDto cardToUpdate) {
+        Long columnId = cardService.update(cardId, cardToUpdate);
+        System.out.println(columnId);
+        System.out.println(cardToUpdate.getColumnId());
```
`logger` 를 잘 선언했지만 `.println()` 을 쓰시는군요...

jihye-woo replied:
앗... 다음부터는 꼭 `logger`를 사용하겠습니다..🥲

jihye-woo replied:
아래와 같이 수정했습니다 휴우..🥲
로그 레벨은 info로 할지 debug로 할지 고민이 많았는데..!
개발 시에 디버깅하기 위한 용도로 로그를 찍는거라 로그 레벨은 debug로 설정했습니다..!! 

```java
 logger.debug("Card to update has column id {}.", columnId);
 logger.debug("Updated card info has column id {}.", cardToUpdate.getColumnId());    
```

wheejuni commented on backend/src/main/java/com/team08/todolist/controller/ApiCardController.java:
```diff
@@ -0,0 +1,63 @@
+        System.out.println(cardToUpdate.getColumnId());
+        if (columnId == cardToUpdate.getColumnId()) {
+            historyService.generateUpdate(cardToUpdate);
+        } else {
```
여기 안에서 `return cardToUpdate` 를 하면, 아래 `else` 는 필요없겠죠?

jihye-woo replied:
else를 없애고, 위 로직을 아래와 같이 메소드로 따로 분리했습니다..!
else가 없으니 훨씬 깔끔하네요👍 감사합니당
```java
private void generateUpdatedCardHistory(Long columnId, CardDto cardToUpdate) {
        if (columnId == cardToUpdate.getColumnId()) {
            historyService.generateUpdate(cardToUpdate);
            return;
        }
        historyService.generateMove(columnId, cardToUpdate);
    }
```

wheejuni commented on backend/src/main/java/com/team08/todolist/controller/ApiCardController.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    @DeleteMapping("/{cardId}")
+    public String delete(@PathVariable Long cardId) {
```
가급적이면 REST API의 응답을 내릴땐, 순수 `String` 보다는 응답 래핑 객체를 만들어 내리는 것을 원칙으로 삼아 주세요.

jihye-woo replied:
말씀하신대로 응답 래핑 용도로 ApiResponse라는 클래스를 아래와 같이 만들었습니다!!
String 타입 뿐만 아니라 응답에 대한 모든 요청을 ApiResponse 객체에 담아서 보내도록 수정하였습니다..!

```java
public class ApiResponse<T> {

    private final T response;

    private final String message;

    private ApiResponse(T response, String message) {
        this.response = response;
        this.message = message;
    }
// ...
}
```
클라이언트로 응답을 보내는 형식에 통일성이 생긴 느낌이라 개인적으로는 기분이 좋네요..😃 
혹시 문제되는 부분이 있으면!! 말씀해주시면 감사드리겠습니당..!😅

wheejuni commented on backend/src/main/java/com/team08/todolist/dto/CardDto.java:
```diff
@@ -0,0 +1,76 @@
+    private Integer position;
+    private Long columnId;
+
+    public CardDto(Long cardId, String title, String content, boolean isApp, String createdTime, Integer position, Long columnId) {
```
생성자에 담을 정보가 굉장히 많네요.
이런 경우 생성자를 `private` 정도로 가리고, 빌더 패턴이나 스태틱 팩토리 메소드 사용을 추천합니다.

jihye-woo replied:
안그래도 생성자 파라미터에 데이터들이 늘어나면서 데이터 순서가 헷갈리기 시작하더라구요..😂

그래서 말씀하신대로 엔티티의 생성자는 `private`으로 설정하고, Card와 CardDto 내부에는 `Builder`를 만들어서 빌더를 통해 객체를 생성할 수 있도록 수정하였습니다!!

```java
 // CardDto
    public static Builder build(Long cardId, Integer position, Long columnId) {
        return new Builder(cardId, position, columnId);
    }

    static public class Builder {
        private Long cardId;
        private String title = "no title";
        private String content = "no contents";
        private boolean isApp = true;
        private LocalDateTime createdDateTime = LocalDateTime.now();
        private Integer position;
        private Long columnId;

        private Builder(Long cardId, Integer position, Long columnId) {
            this.cardId = cardId;
            this.position = position;
            this.columnId = columnId;
        }
   // ...
    }
```
파라미터 순서를 생각하지 않고 데이터를 집어넣어도 된다는게 제일 행복하네요..✨ 감사합니당👍👍
```java
// 빌더를 사용해서 CardDto 객체 생성
public static CardDto of(Card card) {
        return CardDto.build(card.getCid(), card.getPos(), card.getColumnId())
                .title(card.getTitle())
                .content(card.getTitle())
                .isApp(card.isApp())
                .createdDateTime(card.getDate())
                .build();
    }
```

wheejuni commented on backend/src/main/java/com/team08/todolist/dto/CardDto.java:
```diff
@@ -0,0 +1,76 @@
+        return new Card(null, "august17", title, content, createdDateTime, isApp, columnId, position);
+    }
+
+    public static CardDto of(Card card) {
```
👍

wheejuni commented on backend/src/main/java/com/team08/todolist/repository/CardRepository.java:
```diff
@@ -0,0 +1,20 @@
+public interface CardRepository extends CrudRepository<Card, Long> {
+
+    @Override
+    @Query("SELECT CID, TITLE, CONTENT, `DATE`, POS, COLUMNID, AUTHOR, IF(`CARD`.ISAPP, 'true', 'false') as ISAPP from `CARD`")
```
`IF(`CARD`.ISAPP, 'true', 'false')` 이 부분은 불필요하지 않을까요?

jihye-woo replied:
int → boolean 치환하는 로직을 쿼리문에 넣지 않을 경우 아래와 같은 에러가 발생합니다ㅠㅠ
 
>There was an unexpected error (type=Internal Server Error, status=500).
>No converter found capable of converting from type [java.lang.Integer] to type [boolean]
>org.springframework.core.convert.ConverterNotFoundException: No converter found capable of >converting from type [java.lang.Integer] to type [boolean]


DB에서는 boolean 값이 따로 없고 TINYINT 밖에서 없어서 `isApp`의 데이터 타입을 int로 변경하거나 IF()로 쿼리문에서 변환 처리(int ➡️ boolean)하는 작업이 필요한 것 같습니다..ㅠㅠ

wheejuni commented on backend/src/main/java/com/team08/todolist/repository/HistoryRepository.java:
```diff
@@ -0,0 +1,12 @@
+
+public interface HistoryRepository extends CrudRepository<History, Long> {
+
+    @Override
+    List<History> findAll();
```
`.findAll` 은 이미 있는 메소드인데요? 왜 추가적으로 작성하셨나요?

jihye-woo replied:
CrudRepository에서 기본적으로 제공하는 findAll()의 리턴 타입이 `Iterable<T>`인데 리턴 타입을 `List<History>`로 바꾸기 위해서 오버라이드 해주었습니다!!

wheejuni commented on backend/src/main/java/com/team08/todolist/service/CardService.java:
```diff
@@ -0,0 +1,43 @@
+        return cardRepository.findAll();
+    }
+
+    public CardDto delete(Long cardId) {
```
지웠는데 `CardDto` 가 리턴되는 건 좀 어색하지 않을까요?

jihye-woo replied:
delete한 이후 history(로그) 생성을 위해서 삭제된 CardDto 정보를 리턴해주었습니다..!

```java
// ApiCardController
CardDto cardDto = cardService.delete(cardId);
historyService.generateRemove(cardDto);
```

```java
// HistoryService에서 삭제 액션에 대한 History 생성 메소드
 public void generateRemove(CardDto cardDto) {
        History history = new History(null, "august17", cardDto.getTitle(), "remove", null, getColumnType(cardDto.getColumnId()), LocalDateTime.now());
        historyRepository.save(history);
    }
```

jihye-woo replied:
로직에 문제가 있는걸까요..?ㅠㅠ 
다른 방식으로 삭제 액션에 대한 히스토리 생성을 할 수 있도록 다시 한 번 더 생각해보겠습니당...!!🤔

wheejuni commented on backend/src/main/java/com/team08/todolist/service/CardService.java:
```diff
@@ -0,0 +1,43 @@
+        return CardDto.of(card);
+    }
+
+    public Long update(Long cardId, CardDto cardToUpdate) {
```
지금은 도메인이 단순해서 크게 문제가 없었을 것 같은데 `UPDATE` 성 메소드엔 `@Transactional` 을 항상 붙여주세요.
도메인 객체가 조금만 더 복잡해지고, 관계 매핑이 들어가는 부분이 생기게 되면 문제에 부딪힐 수도 있습니다.

jihye-woo replied:
Transactional 처리를 안해주었네요ㅠㅠ 

서비스 레이어에서
1. `UPATE` 성 (INSERT, UPDATE, DELETE에 해당하는 메소드) 메소드의 경우에는 @Transactional을,
2. `SELECT`(단순 조회)를 위한 메소드에는 @Transactional(readOnly=true)를 붙여 주었습니다!

wheejuni commented on backend/src/main/java/com/team08/todolist/service/HistoryService.java:
```diff
@@ -0,0 +1,56 @@
+        historyRepository.save(history);
+    }
+
+    private String getColumnType(Long columnId) {
+        if (columnId == 1L) {
+            return "ToDo";
+        } else if (columnId == 2L) {
+            return "Doing";
+        } else if (columnId == 3L) {
+            return "Done";
+        }
```
별도의 `enum` 으로 관리돼야 하지 않을까요?

jihye-woo replied:
```java
public enum ColumnType {

    TODO("ToDo"),
    DOING("Doing"),
    DONE("Done");
 // ...
}
```

사실 처음에는 enum을 사용하게 되면 column 타입 처리와 관련한 코드들이 지금보다 늘어나게 되어서 불필요하다고 생각을 했었습니다...!😅

근데 다시 생각을 해보니 `ColumnType` enum 클래스로 column 타입 처리와 관련한 책임을 분리해서 서비스 레이어에 불필요한 로직을 제거한다고 생각하니 바로 만들어주어야 겠다는 생각이 들었습니다..!👀 (레퍼런스 : 디온)

덕분에 enum의 효용성에 대해서도 더 생각해보게 되었네용 감사합니당👍👍

wheejuni commented on backend/src/main/resources/banner.txt:
```diff
@@ -0,0 +1,16 @@
+${AnsiColor.BLUE}
+  /$$$$$$                  /$$                  /$$$$$$                                      /$$
+ /$$__  $$                | $$                 /$$__  $$                                    | $$
+| $$  \__/  /$$$$$$   /$$$$$$$  /$$$$$$       | $$  \__/  /$$$$$$  /$$   /$$  /$$$$$$   /$$$$$$$
+| $$       /$$__  $$ /$$__  $$ /$$__  $$      |  $$$$$$  /$$__  $$| $$  | $$ |____  $$ /$$__  $$
+| $$      | $$  \ $$| $$  | $$| $$$$$$$$       \____  $$| $$  \ $$| $$  | $$  /$$$$$$$| $$  | $$
+| $$    $$| $$  | $$| $$  | $$| $$_____/       /$$  \ $$| $$  | $$| $$  | $$ /$$__  $$| $$  | $$
+|  $$$$$$/|  $$$$$$/|  $$$$$$$|  $$$$$$$      |  $$$$$$/|  $$$$$$$|  $$$$$$/|  $$$$$$$|  $$$$$$$
+ \______/  \______/  \_______/ \_______/       \______/  \____  $$ \______/  \_______/ \_______/
+                                                              | $$
+                                                              | $$
+                                                              |__/
```
멋지네요;

wheejuni commented on backend/src/main/resources/application.properties:
```diff
@@ -1,4 +1,7 @@
 spring.datasource.username=team08
-spring.datasource.password=codesquad
\ No newline at end of file
+spring.datasource.password=codesquad
```
아직은 괜찮지만 이걸 숨길 방법도 고민해야겠죠?
공개 레포에 암호 노출은 정말 정말 **위험합니다!**

jihye-woo replied:
ㅇ앗... 그렇군요 아무 생각 없이 넘어갔는데 다시 생각해보니, "비밀번호 여기있습니다~ 여기 보세요~ "이런 느낌이군요..😅
감사합니다!!👍👍

wheejuni replied:
좋은 느낌이네요 ㅋㅋ

wheejuni reviewed (changes requested):
수고 많으셨습니다. 👍 
졸린 눈을 부비며 프로젝트를 진행했을 여러분의 모습을 떠올려보니 코드를 더 조심스럽게 읽게 되었습니다.

손봤으면 하는 부분은 모두 피드백으로 남겨 두었습니다.
큰 수정이 필요할 것 같진 않습니다. 💯 

수고하셨습니다~

jihye-woo committed:
refactor: Rename field as a DB columns name (#25)

- 수정된 Card 테이블 컬럼명과 매핑되도록 Card 엔티티 클래스 필드명과 getter 메소드명 수정

jihye-woo committed:
feat: Add ApiResponse for wrapping the response data (#25)

- 클라이언트에게 응답을 보내야할 때 ApiResponse(응답 래핑 객체)로 통일해서 전달

jihye-woo committed:
feat: Add inner class Builder at Card, CardDto class (#25)

- Card와 CardDto 객체 생성 시 빌더 패턴 활용

jihye-woo committed:
feat: Add inner class Builder at Card, CardDto class (#25)

- Card와 CardDto 객체 생성 시 빌더 패턴 활용

jihye-woo committed:
feat: Add @Transactional at service layer method (#25)

- CardService 메소드에 Transactional 어노테이션 추가하여 트렌젝션 보장

jihye-woo committed:
feat: Add ColumnType enum class (#25)

- ColumnType enum 클래스로 컬럼 유형을 분류
- Column 타입 정보를 처리하는 책임을 ColumnType 클래스로 이동

jihye-woo committed:
refactor: Change response type as ApiResponse (#25)

- 각 컨트롤러 메소드의 응답 관련 정보 ApiResponse 객체에 담아서 리턴

jihye-woo committed:
fix: Change log information for columnId (#25)

- 인자로 받아온 CardDto의 columnId 정보를 logger로 출력

jihye-woo committed:
feat: Add @Transactional at HistoryService (#25)

- HistoryService의 모든 메소드에 Transactional 혹은 Transactional readOnly 처리
- CardService에서 SELECT 성 메소드를 위해서 @Transactional(ReadOnly=true) 설정
